hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheRecordStore.java

```diff
@@ -403,6 +426,68 @@ protected R processExpiredEntry(Data key, R record, long expiryTime, long now, S
     }
 
     protected void onProcessExpiredEntry(Data key, R record, long expiryTime, long now, String source, String origin) {
+        accumulateOrSendExpiredKeysToBackup(key, record);
+    }
+
+    protected void accumulateOrSendExpiredKeysToBackup(Data key, R record) {
+        if (cacheConfig.getBackupCount() == 0) {
+            return;
+        }
+        if (key != null && record != null) {
+            this.expiredKeys.offer(new ExpiredKey(toHeapData(key), record.getCreationTime()));
+        }
+
+        sendBackupExpirations(true);
+    }
+
+    @Override
+    public boolean isExpirable() {
+        return hasEntryWithExpiration || getConfig().getExpiryPolicyFactory() != null;
+    }
+
+    public void sendBackupExpirations(boolean checkIfReachedBatch) {
+        InvalidationQueue<ExpiredKey> invalidationQueue = this.expiredKeys;
+
+        int size = invalidationQueue.size();
+        if (size == 0 || checkIfReachedBatch && size < MAX_EXPIRED_KEY_COUNT_IN_BATCH) {
+            return;
+        }
+
+        if (!invalidationQueue.tryAcquire()) {
+            return;
+        }
+
+        Collection<ExpiredKey> expiredKeys = new ArrayList<ExpiredKey>();
```

Comment:
Can't we replace this whole loop thingy with:
```java
    Collection<ExpiredKey> expiredKeys = new ArrayList<ExpiredKey>(invalidationQueue);
    invalidationQueue.release();
```
I think this will perform much better, since the `ArrayList` will never need to be resized. If the code can be simplified, the NPath complexity suppression can also be removed.

If we need this complex loop for some reason, you could still pull this out in a method like
```java
Collection<ExpiredKey> expiredKeys = getExpiredKeys(invalidationQueue);
if (expiredKeys.size() == 0) {
    return;
}
```
which will also get rid of the NPath complexity.